### PR TITLE
feat: Improve BMR test coverage, BMR refactoring #399

### PR DESCRIPTION
**Pull request by andrii-kl**

### CODE REVIEW

**Comment by CyrusVorwald-ICON** [comment]
This is a very large number of changes to review. Can this be split it up into many smaller reviews?

**Comment by andrii-kl** [comment]
> This is a very large number of changes to review. Can this be split it up into many smaller reviews?

Unfortunately the receiver file has high coupling code and a lot of connection with different components. 
I tried to minimize changes scope, so current PR doesn't have optional changes.

In the next steps I there is a plan to remove reciver_core, It will require some additional refactoring but should increase total code coverage a lot.

**Comment by bbist** [comment]
Hi @andrii-kl,
There are a lot of changes. The full review might take some time. What I'll do is, make some changes and refactoring to fix minor issues and push the changes. Will need to run the end-to-end test before merging into main. If it works out, we can merge and start making modifications on top of that. I hope you'll be able to take up and continue this after SNOW integration. Even though the coverage might be high, it still needs to cover edge cases, that we can discuss once you start working on it next time.

**Review by bbist — CHANGES_REQUESTED** [review]

File: cmd/iconbridge/chain/icon/client.go [hunk 1]
```diff
@@ -198,7 +228,7 @@ func (c *Client) WaitForResults(ctx context.Context, thp *TransactionHashParam)
 				switch re := err.(type) {
 				case *jsonrpc.Error:
 					switch re.Code {
-					case JsonrpcErrorCodePending, JsonrpcErrorCodeExecuting, JsonrpcErrorCodeNotFound:
+					case types.JsonrpcErrorCodePending, types.JsonrpcErrorCodeNotFound:
```
Comment (manishbista28) [inline cmd/iconbridge/chain/icon/client.go:231]: types.JsonrpcErrorCodeExecuting seems to be missing.

File: cmd/iconbridge/chain/icon/register_relay.go [hunk 2]
```diff
@@ -1,8 +0,0 @@
-package icon
```
[the remark this answers is not in the document]
Comment (andrii-kl) [reply cmd/iconbridge/chain/icon/register_relay.go]: I got it, I didn't find any use of this file, I thought it was unused.

**Comment by bbist** [comment]
@andrii-kl, @manishbista28,

Are there any other issues that we want to push in current release? If there are, we need to merge them before this PR. Only then we should merge this PR.
I'd like to make this PR part of next release, because even though the test coverage percentage is high the actual tests don't include all critical paths. So, before releasing this PR we'd want to resolve all those issues.

**Comment by bbist** [comment]
> @andrii-kl, @manishbista28, @pragyanshrestha-ibriz,
> 
> Are there any other issues that we want to push in current release? If there are, we need to merge them before this PR. Only then we should merge this PR. I'd like to make this PR part of next release, because even though the test coverage percentage is high the actual tests don't include all critical paths. So, before releasing this PR we'd want to resolve all those issues.

I'm approving this change as e2e test seems to be running fine. Any changes after this PR is merge into main should be made part of next release.

**Review by bbist — APPROVED** [review]

**Comment by andrii-kl** [comment]
https://github.com/icon-project/icon-bridge/issues/614

**Pull request merged**